### ISSUE TITLE
[3.x] Out of bounds access in Vector2 causing crash

### ISSUE THREAD

**Issue by lawnjelly** [godotengine/godot#58423]
### Godot version

3.x latest (22 Feb 2022) 3e0f380

### System information

Linux Mint 20, HD Graphics 630

### Issue description

Crashes loading up the 3D Kinematic Character Godot demo project in the editor:

(this is before Calinou's recent adjustment to the project BTW just in case that made a difference)
```
1  GIProbeGizmoPlugin::redraw    spatial_editor_gizmos.cpp 2831 0x4d50b40 
2  EditorSpatialGizmo::redraw    spatial_editor_gizmos.cpp 114  0x4d28c8c 
3  Spatial::set_gizmo            spatial.cpp               498  0x59ee283 
4  SpatialEditor::_request_gizmo spatial_editor_plugin.cpp 6386 0x5274304 
5  MethodBind1<Object *>::call   method_bind.gen.inc       775  0x374fe20 
6  Object::call                  object.cpp                918  0x667ef1c 
7  MessageQueue::_call_function  message_queue.cpp         241  0x6672e39 
8  MessageQueue::flush           message_queue.cpp         284  0x667313f 
9  SceneTree::idle               scene_tree.cpp            579  0x55035cf 
10 Main::iteration               main.cpp                  2269 0x313c714 
11 OS_X11::run                   os_x11.cpp                3722 0x30f274e 
12 main                          godot_x11.cpp             55   0x30dd0b6 
```
I think this is a recent regression, it is working fine in 3.5 beta 1 and 3.4.3 RC 2. Possibly one of the cherry picks?
Trying to bisect it now.

I've tried a clean and rebuild etc just in case it was a corruption in my build but that hasn't solved it.

### Steps to reproduce

See above.

## Bisecting
Feb 15th broken
Feb 9th S3TC commit broken

Is either-
Feb 9th vector2 brackets
Feb 9th infer variable types on release

Feb 9th dirty flag - works
Feb 8th PR 56093 works

(Bizarrely, downloading a windows artifact build from Feb 15th works under wine, so it's possible it's linux only or some kind of corruption related bug)

**Comment by lawnjelly** [comment]
Bisected this down to #57844 .

I can immediately see that this could could an out of bounds read if the calling code is incorrect, whereas the previous version would prevent such out of bounds access.

So I don't think the PR is necessarily "wrong", but we need to track down where this is being called. We can perhaps put a debug check on the accessor to identify this problem, and flag it.

**Comment by Calinou** [comment]
Related to https://github.com/godotengine/godot/issues/58284.

**Comment by akien-mga** [comment]
> Related to #58284. I thought this affected only `master`?

That would also have been caused by #57843 which exposes a GIProbe gizmo bug, which #58428 fixes for `3.x` (still needs a fix in `master`).

**Comment by akien-mga** [comment]
The crash was fixed by #58428. A dev check was added in #58427 to warn users about it, but in release builds this stays unchecked and would crash like any other out-of-bounds read. This is expected as this is API misuse and guarding against it would have a strong performance impact.
